src/memory-analyzer/gdb_api.cpp

```diff
@@ -0,0 +1,266 @@
+// Copyright 2018 Author: Malte Mues <mail.mues@gmail.com>
+#ifdef __linux__
+#include <cctype>
+#include <cerrno>
+#include <cstdio>
+#include <cstring>
+#include <regex>
+
+#include "gdb_api.h"
+#include <goto-programs/goto_model.h>
+
+gdb_apit::gdb_apit(const char *arg)
+  : binary_name(arg), buffer_position(0), last_read_size(0)
+{
+  memset(buffer, 0, MAX_READ_SIZE_GDB_BUFFER);
+}
+
+int gdb_apit::create_gdb_process()
+{
+  if(pipe(pipe_input) == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create pipe for stdin!");
+  }
+  if(pipe(pipe_output) == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create pipe for stdout!");
+  }
+
+  gdb_process = fork();
+  if(gdb_process == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create gdb process.");
+  }
+  if(gdb_process == 0)
+  {
+    // child process
+    close(pipe_input[1]);
+    close(pipe_output[0]);
+    dup2(pipe_input[0], STDIN_FILENO);
+    dup2(pipe_output[1], STDOUT_FILENO);
+    dup2(pipe_output[1], STDERR_FILENO);
+    dprintf(pipe_output[1], "binary name: %s\n", binary_name);
+    char *arg[] = {
+      const_cast<char *>("gdb"), const_cast<char *>(binary_name), NULL};
+
+    dprintf(pipe_output[1], "Loading gdb...\n");
+    execvp("gdb", arg);
+
+    // Only reachable, if execvp failed
+    int errno_value = errno;
+    dprintf(pipe_output[1], "errno in child: %s\n", strerror(errno_value));
+  }
+  else
+  {
+    // parent process
+    close(pipe_input[0]);
+    close(pipe_output[1]);
+    write_to_gdb("set max-value-size unlimited\n");
+  }
+  return 0;
+}
+
+void gdb_apit::terminate_gdb_process()
+{
+  close(pipe_input[0]);
+  close(pipe_input[1]);
+}
+
+void gdb_apit::write_to_gdb(const std::string &command)
+{
+  if(
+    write(pipe_input[1], command.c_str(), command.length()) !=
+    (ssize_t)command.length())
+  {
+    throw gdb_interaction_exceptiont("Could not write a command to gdb");
+  }
+}
+
+std::string gdb_apit::read_next_line()
+{
+  char token;
+  std::string line;
+  do
+  {
+    while(buffer_position >= last_read_size)
+    {
+      read_next_buffer_chunc();
+    }
+    token = buffer[buffer_position];
+    line += token;
+    ++buffer_position;
+  } while(token != '\n');
+  return line;
+}
+
+void gdb_apit::read_next_buffer_chunc()
+{
+  memset(buffer, 0, last_read_size);
+  const auto read_size =
+    read(pipe_output[0], &buffer, MAX_READ_SIZE_GDB_BUFFER);
+  if(read_size < 0)
+  {
+    throw gdb_interaction_exceptiont("Error reading from gdb_process");
+  }
+  last_read_size = read_size;
+  buffer_position = 0;
+}
+
+void gdb_apit::run_gdb_from_core(const std::string &corefile)
```

Comment:
Should the public interfaces to this API (i.e. the two 'run...' methods) actually check that gdb is running? At the moment I think they will block in the first write_to_gdb call if the user either hasn't called create_gdb_process, or has called terminate_gdb_process?

Reply:
Yes, this is now checked via a precondition.